File: js/selfcheck.js
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');

const files = {
  html: path.join(ROOT, 'index.html'),
  css: path.join(ROOT, 'css', 'ui.css'),
  main: path.join(ROOT, 'js', 'main.js'),
  story: path.join(ROOT, 'data', 'story.js'),
  chap1: path.join(ROOT, 'data', 'chapters', 'chap1.js'),
  chap2: path.join(ROOT, 'data', 'chapters', 'chap2.js'),
  drinks: path.join(ROOT, 'data', 'drinks.js'),
  state: path.join(ROOT, 'js', 'state.js'),
  validator: path.join(ROOT, 'tools', 'validate_story.js'),
  balanceReport: path.join(ROOT, 'tools', 'balance_report.js'),
  basicCheck: path.join(ROOT, 'tests', 'basic_check.js')
};

function fail(msg) {
  console.error(`[FAIL] ${msg}`);
  process.exit(1);
}

function ok(msg) {
  console.log(`[OK] ${msg}`);
}

function read(file) {
  if (!fs.existsSync(file)) fail(`missing file: ${path.relative(ROOT, file)}`);
  return fs.readFileSync(file, 'utf8');
}

const html = read(files.html);
const main = read(files.main);
const css = read(files.css);
const story = read(files.story);
const chap1 = read(files.chap1);
const chap2 = read(files.chap2);
const chapterStory = `${chap1}\n${chap2}`;
const drinks = read(files.drinks);
const state = read(files.state);
read(files.validator);
const balanceReport = read(files.balanceReport);
const basicCheck = read(files.basicCheck);

const sceneCount = (chapterStory.match(/\bs\d{2}[A-C]?\s*:\s*{/g) || []).length;
if (sceneCount < 12) fail(`scene count ${sceneCount} < 12`);
ok(`scene count: ${sceneCount}`);

const endings = ['结局A【', '结局B【', '结局C【'];
if (!endings.every((item) => main.includes(item))) fail('A/B/C endings missing');
ok('A/B/C endings exist');

if (!story.includes('chapter1Scenes') || !story.includes('chapter2Scenes')) fail('chapter merge import missing in story.js');
if (!story.includes('...chapter1Scenes') || !story.includes('...chapter2Scenes')) fail('chapter spread merge missing in story.js');
ok('chapter merge structure exists');

const branchSceneCount = (chap2.match(/\bs1[1-5]\s*:\s*{/g) || []).length;
if (branchSceneCount < 4) fail(`side branch scenes ${branchSceneCount} < 4`);
ok(`side branch scenes: ${branchSceneCount}`);

for (const token of ["flagsAll: ['corpDeal', 'truthLeakDraft']", "relAtLeast: { name: 'hacker', val: 1 }", "setFlags: ['ghostHandshake']"]) {
  if (!chap2.includes(token)) fail(`missing side-branch token: ${token}`);
}
ok('side-branch trigger and completion tokens exist');

for (const token of ['st.flags.ghostHandshake', 'sideQuestSuffix', '灰匣注释']) {
  if (!(`${chap2}\n${main}`).includes(token)) fail(`missing ending variation token: ${token}`);
}
ok('ending variation tokens exist');

const schemaMatch = main.match(/const\s+SAVE_SCHEMA_VERSION\s*=\s*(\d+)/);
if (!schemaMatch || Number(schemaMatch[1]) < 1) fail('invalid SAVE_SCHEMA_VERSION');
ok(`schemaVersion: ${schemaMatch[1]}`);

const drinkCount = (drinks.split('export const EXTRAS')[0].match(/id:\s*'[^']+'/g) || []).length;
if (drinkCount < 8) fail(`drink definitions ${drinkCount} < 8`);
ok(`drink definitions: ${drinkCount}`);

const importantFlagMatch = chapterStory.match(/const\s+importantFlags\s*=\s*\[([\s\S]*?)\]/);
const fallbackFlags = ['corpDeal', 'hackerTrust', 'policeWarrant', 'memoryTape', 'barShielded', 'truthLeakDraft', 'ghostHandshake'];
const flags = importantFlagMatch
  ? [...importantFlagMatch[1].matchAll(/'([^']+)'/g)].map((m) => m[1])
  : fallbackFlags.filter((f) => chapterStory.includes(f) || story.includes(f));
if (flags.length < 6) fail(`important flags ${flags.length} < 6`);
ok(`important flags: ${flags.length}`);

const requiredKeys = ['sceneId', 'tendencies', 'flags', 'dialogueHistory', 'audioSettings', 'orderHistory'];
for (const key of requiredKeys) {
  if (!main.includes(key)) fail(`missing save key token: ${key}`);
}
ok('save payload tokens exist');

for (const token of ['routeForecast', 'routeDiagnosisList', 'normalizeEffectVector', 'tendencyDelta', '终章预测：代号']) {
  if (!(`${html}\n${main}`).includes(token)) fail(`missing diagnosis token: ${token}`);
}
ok('route diagnosis + forecast tokens exist');

if (!balanceReport.includes('choice balance report') || !balanceReport.includes('imbalance hints')) fail('balance report script markers missing');
ok('balance report script markers exist');

const keyboardTokens = ["event.key === 'Enter'", "['1', '2', '3']", "event.key.toLowerCase() === 's'", "event.key.toLowerCase() === 'l'"];
for (const token of keyboardTokens) {
  if (!main.includes(token)) fail(`missing keyboard token: ${token}`);
}
ok('keyboard shortcuts exist');

if (!html.includes('aria-label')) fail('aria-label not found in index.html');
ok('aria-label exists');


const layoutTokens = [
  'overflow: hidden',
  'height: 100%',
  'height: calc(var(--vh) * 100 - 14px)',
  '.story.story-clamped',
  '.choices',
  '.story-modal',
  'max-height: calc(100dvh - 24px)'
];
for (const token of layoutTokens) {
  if (!css.includes(token)) fail(`missing responsive layout token in css/ui.css: ${token}`);
}
ok('responsive anti-overflow tokens exist in css/ui.css');

for (const token of ['updateViewportHeightVar', 'reportNoScroll', 'togglePanelWithLayoutCheck', 'expandStoryBtn', 'storyModal']) {
  if (!main.includes(token)) fail(`missing runtime layout check token: ${token}`);
}
ok('runtime layout self-check helpers exist');

for (const token of ['scrollHeight', 'scrollWidth', '390x844']) {
  if (!basicCheck.includes(token)) fail(`missing no-scroll test token: ${token}`);
}
ok('tests/basic_check.js no-scroll checks exist');

if (!state.includes('hasCondition')) fail('state condition checker missing');
ok('state condition checker exists');

if (!story.includes('STORY_SCHEMA') || !story.includes('STORY_FLAG_WHITELIST')) fail('story schema exports missing');
ok('story schema exports exist');

console.log('\nSelf-check passed.');

File: scripts/selfcheck.py
#!/usr/bin/env python3
"""Repository self-check for neon-tape-vn."""

from __future__ import annotations

import pathlib
import re
import sys

ROOT = pathlib.Path(__file__).resolve().parents[1]
INDEX = ROOT / "index.html"
MAIN_JS = ROOT / "js" / "main.js"
DATA_JS = ROOT / "js" / "data.js"


def fail(message: str) -> None:
    print(f"[FAIL] {message}")
    sys.exit(1)


def ok(message: str) -> None:
    print(f"[OK] {message}")


def main() -> None:
    for path in [INDEX, MAIN_JS, DATA_JS]:
        if not path.exists():
            fail(f"missing required file: {path.relative_to(ROOT)}")

    html = INDEX.read_text(encoding="utf-8")
    main_js = MAIN_JS.read_text(encoding="utf-8")
    data_js = DATA_JS.read_text(encoding="utf-8")

    required_strings = [
        "NEON TAPE_017",
        "剧情日志（可回看）",
        "绫濑雾音｜企业联络官",
        "烬线｜街头黑客",
        "韩铬｜义体警探",
        "路线锁定",
        "结局A【玻璃停火】",
        "结局B【霓虹燃烧】",
        "结局C【磁带群星】",
        "Tape Archive / 存档总线",
        "schemaVersion",
        "saveTransferText",
    ]
    merged = html + "\n" + main_js + "\n" + data_js
    for token in required_strings:
        if token not in merged:
            fail(f"missing required string: {token}")
    ok("key strings present")

    scene_count = len(re.findall(r"\bs\d{2}[A-C]?\s*:\s*{", data_js))
    if scene_count < 13:
        fail(f"scene count too low: {scene_count} (expected >= 13)")
    ok(f"scene count >= 13 ({scene_count})")

    endings = len(re.findall(r"结局[A-C]【", main_js))
    if endings < 3:
        fail(f"ending markers too low: {endings} (expected >= 3)")
    ok(f"ending markers >= 3 ({endings})")

    route_finals = len(re.findall(r"s10[A-C]", data_js))
    if route_finals < 3:
        fail(f"route specific final scenes too low: {route_finals} (expected >= 3)")
    ok(f"route specific final scenes >= 3 ({route_finals})")

    keys = sorted(set(re.findall(r"neonTape_[a-zA-Z0-9_]*", merged)))
    if not keys or 'neonTape_' not in keys:
        fail("no save key markers found")
    ok(f"save key markers present ({', '.join(keys)})")

    if "data-save=\"slot1\"" not in html or "data-save=\"slot2\"" not in html or "data-save=\"slot3\"" not in html:
        fail("manual save slots < 3")
    ok("manual save slots >= 3")

    for token in ["routeLock", "choiceHistory", "AUTO_SLOT", "sceneId", "unlockedEndings", "bgmEnabled", "bgmVolume", "SAVE_SCHEMA_VERSION"]:
        if token not in main_js:
            fail(f"missing save payload key indicator: {token}")
    ok("save payload includes route and review keys")

    for token in ["data-export=\"slot1\"", "data-import=\"slot1\"", "data-export=\"auto\"", "data-import=\"auto\""]:
        if token not in html:
            fail(f"missing import/export control: {token}")
    ok("import/export controls present")

    print("\nSelf-check passed.")


if __name__ == "__main__":
    main()
